dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -692,48 +694,97 @@ DecodeDetail ExchangeReceiverBase<RPCContext>::decodeChunks(
 
     // Record total packet size even if fine grained shuffle is enabled.
     detail.packet_bytes = packet.ByteSizeLong();
-
+    detail.produce_block_flag = false;
     for (const String * chunk : recv_msg->chunks)
     {
-        Block block = CHBlockChunkCodec::decode(*chunk, header);
-        detail.rows += block.rows();
-        if (unlikely(block.rows() == 0))
+        auto result = decoder_ptr->decodeAndSquash(*chunk);
+        if (!result)
             continue;
-        block_queue.push(std::move(block));
+        detail.rows += result->rows();
+        if likely(result->rows() > 0)
+        {
+            detail.produce_block_flag = true;
+            block_queue.push(std::move(result.value()));
+        }
     }
     return detail;
 }
 
 template <typename RPCContext>
-ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::nextResult(std::queue<Block> & block_queue, const Block & header, size_t stream_id)
+std::vector<ExchangeReceiverResult> ExchangeReceiverBase<RPCContext>::nextResult(
+    std::queue<Block> & block_queue,
+    const Block & header,
+    size_t stream_id,
+    std::unique_ptr<IChunkDecodeAndSquash> & decoder_ptr)
 {
+    std::vector<ExchangeReceiverResult> results;
     if (unlikely(stream_id >= msg_channels.size()))
     {
         LOG_ERROR(exc_log, "stream_id out of range, stream_id: {}, total_stream_count: {}", stream_id, msg_channels.size());
-        return ExchangeReceiverResult::newError(0, "", "stream_id out of range");
+        results.push_back(ExchangeReceiverResult::newError(0, "", "stream_id out of range"));
+        return results;
     }
-    std::shared_ptr<ReceivedMessage> recv_msg;
-    if (msg_channels[stream_id]->pop(recv_msg) != MPMCQueueResult::OK)
+
+    while (true)
```

Comment:
It looks like both `while(true)` and `std::vector<ExchangeReceiverResult>` are unnecessary, we can do this in `TiRemoteBlockInputStream::fetchRemoteResult`

Reply:
such as
```
auto result = remote_reader->nextResult(block_queue, sample_block, stream_id);
if (result.meet_error)
    throw Exception()
if (result.await)
    continue;
```

Reply:
+1

Reply:
Done, introduced the rowsInfoOnly flag to determine if it's the result previous to EOF.